firedrake/preconditioners/pmg.py

```diff
@@ -274,55 +303,250 @@ def __init__(self, Vf, Vc, Vf_bcs, Vc_bcs):
+
+        Important notes:
+        The input data in x is destroyed in the process.
+        Need to allocate nel*max(mx, nx)*max(my, ny)*max(mz, nz) memory for both x and y.
+        */
+
+        int m,n,k,s,p,lda;
+        char TA1, TA2, TA3;
+        char tran='T', notr='N';
+        double zero=0.0E0, one=1.0E0;
+
+        if(tflag>0){
+           TA1 = tran;
+           TA2 = notr;
+        }else{
+           TA1 = notr;
+           TA2 = tran;
+        }
+        TA3 = TA2;
+
+        m = mx;  k = nx;  n = ny*nz*nel;
+        lda = (tflag>0)? nx : mx;
+        dgemm_(&TA1, &notr, &m,&n,&k, &one, A1,&lda, x,&k, &zero, y,&m);
+
+        p = 0;  s = 0;
+        m = mx;  k = ny;  n = my;
+        lda = (tflag>0)? ny : my;
+        for(int i=0; i<nz*nel; i++){
+           dgemm_(&notr, &TA2, &m,&n,&k, &one, y+p,&m, A2,&lda, &zero, x+s,&m);
+           p += m*k;
+           s += m*n;
+        }
+
+        p = 0;  s = 0;
+        m = mx*my;  k = nz;  n = mz;
+        lda = (tflag>0)? nz : mz;
+        for(int i=0; i<nel; i++){
+           dgemm_(&notr, &TA3, &m,&n,&k, &one, x+p,&m, A3,&lda, &zero, y+s,&m);
+           p += m*k;
+           s += m*n;
+        }
+        return;
+        }
+        """
 
+        # UFL elements order the component DoFs related to the same node contiguously.
+        # We transpose before and after the multiplcation times J to have each component
+        # stored contiguously as a scalar field, thus reducing the number of dgemm calls.
 
-void restriction(double *restrict Rc, const double *restrict Rf, const double *restrict w)
-{{
-    double Afc[{dimf}*{dimc}] = {{0}};
-    expression_kernel(Afc);
-    for (int32_t i = 0; i < {dimf}; i++)
-       for (int32_t j = 0; j < {dimc}; j++)
-           Rc[j] += Afc[i*{dimc} + j] * Rf[i] / w[i];
-}}
-"""
+        prolong_code = f"""
+        {kronmxv_code}
 
-        self.restrict_kernel = op2.Kernel(self.restrict_code, "restriction")
-        self.mesh = Vf.mesh()
+        void prolongation(double *restrict y, const double *restrict x){{
+            double JX[{mx}*{nx}] = {{ {JX} }};
+            double t0[{lwork}], t1[{lwork}];
+            double one=1.0E0;
+
+            for(int j=0; j<{nxyz}; j++)
+                for(int i=0; i<{nscal}; i++)
+                    t0[j + {nxyz}*i] = x[i + {nscal}*j];
+
+            kronmxv(0, {mx},{my},{mz}, {nx},{ny},{nz}, {nscal}, JX,JX,{JZ}, t0,t1);
+
+            for(int j=0; j<{mxyz}; j++)
+                for(int i=0; i<{nscal}; i++)
+                   y[i + {nscal}*j] = t1[j + {mxyz}*i];
+            return;
+        }}
+        """
+
+        restrict_code = f"""
+        {kronmxv_code}
 
+        void restriction(double *restrict y, const double *restrict x,
+        const double *restrict w){{
+            double JX[{mx}*{nx}] = {{ {JX} }};
+            double t0[{lwork}], t1[{lwork}];
+            double one=1.0E0;
+
+            for(int j=0; j<{mxyz}; j++)
+                for(int i=0; i<{nscal}; i++)
```

Comment:
How big do these loop extents get? If these matrices start falling out of L2 cache, it might be worth loop tiling (https://docs.wence.uk/07-transpose.html)

Reply:
I don't expect `nscal` being greater than 9. I was thinking that we could even unroll the inner loop.

Reply:
Prefer using hints to the compiler than manual unrolling, although these days they're pretty good about doing small extent unrolling. I guess I was thinking of the `mxyz` loop too.

Reply:
`mxyz` = (N+1)^ndim. For my purposes, the degree N could be as large as 32.

Reply:
OK, so loop tiling is likely beneficial at the upper end, but I suspect that we should just leave a comment noting that rather than making this code here more complicated.